### PR TITLE
[Upgrade] Alpha TestNet v0.0.11

### PR DESCRIPTION
**Pull request by okdas**
## Summary

An upgrade for Alpha TestNet from v0.0.10 to v0.0.11.

## Issue

Beta TestNet has been launched using v0.0.11 (release candidate), and we need to upgrade alpha so both networks use the same version.

## Type of change

Chain upgrade.

## Testing

- [] An upgrade is tested by upgrading a network provisioned by the old version (v0.0.10) to the new version (v0.0.11)

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [ ] I have commented my code
- [ ] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

### CODE REVIEW

File: app/upgrades/v0.0.11.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+			sessionParams.NumSuppliersPerSession = uint64(15)
+			err = keepers.SessionKeeper.SetParams(ctx, sessionParams)
+			if err != nil {
+				return err
+			}
+
+			// Set tokenomics params. The values are based on default values for LocalNet.
+			// Validate with: `poktrolld q tokenomics params --node=https://testnet-validated-validator-rpc.poktroll.com/`
+			tokenomicsParams := keepers.TokenomicsKeeper.GetParams(ctx)
+			tokenomicsParams.MintAllocationPercentages = tokenomicstypes.MintAllocationPercentages{
```
Comment (okdas) [inline app/upgrades/v0.0.11.go:43]: These are the same values as Beta TestNet btw.

**Comment by okdas** [comment]
Upgrade has been tested and validated locally:

```
5:39PM INF applying upgrade "v0.0.11" at height: 20 module=x/upgrade
5:39PM INF Starting v0.0.11 upgrade handler module=server
5:39PM INF Starting parameter updates for v0.0.11 module=server
5:39PM INF Current session params module=server params={"num_suppliers_per_session":0}
5:39PM INF Successfully updated session params module=server new_params={"num_suppliers_per_session":15}
5:39PM INF Current tokenomics params module=server params={"dao_reward_address":"","mint_allocation_proposer":{"application":0,"dao":0,"proposer":0,"source_owner":0,"supplier":0}}
5:39PM INF Successfully updated tokenomics params module=server new_params={"dao_reward_address":"pokt1r6ja6rz6rpae58njfrsgs5n5sp3r36r2q9j04h","mint_allocation_proposer":{"application":0,"dao":0.1,"proposer":0.05,"source_owner":0.15,"supplier":0.7}}
5:39PM INF Starting authorization updates for v0.0.11 module=server
5:39PM INF Granting authorization module=server
5:39PM INF Successfully granted authorization module=server
5:39PM INF Running module migrations module=server
```

**Comment by okdas** [comment]
This is interesting.

The [diff between versions](https://github.com/pokt-network/poktroll/compare/v0.0.10...v0.0.11-rc) shows we need to add a new authorization, but we already have it on Alpha TestNet:
```bash
poktrolld q authz grants-by-granter pokt10d07y265gmmuvt4z0w9aw880jnsr700j8yv32t --node=https://shannon-testnet-grove-seed-rpc.alpha.poktroll.com
```


```yaml
- authorization:
    type: cosmos-sdk/GenericAuthorization
    value:
      msg: /poktroll.session.MsgUpdateParam
  expiration: "2500-01-01T00:00:00Z"
  grantee: pokt1r6ja6rz6rpae58njfrsgs5n5sp3r36r2q9j04h
  granter: pokt10d07y265gmmuvt4z0w9aw880jnsr700j8yv32t
```

**Comment by okdas** [comment]
Tested again after authorization code removed:

```

5:58PM INF ABCI Handshake App Info hash=48BAFBE9874B2A7625D9834A983AB231C98523767859E5208A277F8518CF66C0 height=19 module=consensus protocol-version=0 software-version=0.0.11-rc-bc2a3c44
5:58PM INF ABCI Replay Blocks appHeight=19 module=consensus stateHeight=19 storeHeight=20
5:58PM INF Replay last block using real app module=consensus
5:58PM INF applying upgrade "v0.0.11" at height: 20 module=x/upgrade
5:58PM INF Starting v0.0.11 upgrade handler module=server
5:58PM INF Starting parameter updates for v0.0.11 module=server
5:58PM INF Current session params module=server params={"num_suppliers_per_session":0}
5:58PM INF Successfully updated session params module=server new_params={"num_suppliers_per_session":15}
5:58PM INF Current tokenomics params module=server params={"dao_reward_address":"","mint_allocation_proposer":{"application":0,"dao":0,"proposer":0,"source_owner":0,"supplier":0}}
5:58PM INF Successfully updated tokenomics params module=server new_params={"dao_reward_address":"pokt1r6ja6rz6rpae58njfrsgs5n5sp3r36r2q9j04h","mint_allocation_proposer":{"application":0,"dao":0.1,"proposer":0.05,"source_owner":0.15,"supplier":0.7}}
5:58PM INF Running module migrations module=server
5:58PM INF adding a new module: 06-solomachine module=server
5:58PM INF adding a new module: 07-tendermint module=server
```

I think we're good to go. I'll wait for a review, before I cut a new version.

File: app/upgrades/v0.0.11.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	PlanName: "v0.0.11",
+	CreateUpgradeHandler: func(mm *module.Manager,
+		keepers *keepers.Keepers,
+		configurator module.Configurator,
+	) upgradetypes.UpgradeHandler {
+
+		// Adds new parameters using ignite's config.yml as a reference. Assuming we don't need any other parameters.
+		// https://github.com/pokt-network/poktroll/compare/v0.0.10...v0.0.11-rc
+		applyNewParameters := func(ctx context.Context) (err error) {
+			logger := cosmosTypes.UnwrapSDKContext(ctx).Logger()
```
Comment (okdas) [inline app/upgrades/v0.0.11.go:28]: thanks @bryanchriswhite for the hint!

File: app/upgrades/v0.0.11.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+//   - After: v0.0.11
+var Upgrade_0_0_11 = Upgrade{
+	PlanName: "v0.0.11",
+	CreateUpgradeHandler: func(mm *module.Manager,
+		keepers *keepers.Keepers,
+		configurator module.Configurator,
+	) upgradetypes.UpgradeHandler {
+
+		// Adds new parameters using ignite's config.yml as a reference. Assuming we don't need any other parameters.
+		// https://github.com/pokt-network/poktroll/compare/v0.0.10...v0.0.11-rc
```
Comment (Olshansk) [inline app/upgrades/v0.0.11.go:26]: should this be `-rc` or just `0.1`?

File: app/upgrades/v0.0.11.go (hunk 1 above)
Comment (Olshansk) [reply app/upgrades/v0.0.11.go:47]: #PUC?

File: app/upgrades/v0.0.11.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+			tokenomicsParams := keepers.TokenomicsKeeper.GetParams(ctx)
+			logger.Info("Current tokenomics params", "params", tokenomicsParams)
+			tokenomicsParams.MintAllocationPercentages = tokenomicstypes.MintAllocationPercentages{
+				Dao:         0.1,
+				Proposer:    0.05,
+				Supplier:    0.7,
+				SourceOwner: 0.15,
+				Application: 0.0,
+			}
+			tokenomicsParams.DaoRewardAddress = AlphaTestNetPnfAddress
```
Comment (Olshansk) [inline app/upgrades/v0.0.11.go:54]: Should this be `AlphaTestNetPnfAddress` or `BetaTestNetPnfAddress`?

File: app/upgrades/v0.0.11.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+				logger.Error("Failed to set tokenomics params", "error", err)
+				return err
+			}
+			logger.Info("Successfully updated tokenomics params", "new_params", tokenomicsParams)
+
+			return
+		}
+
+		// The diff shows that the only new authz authorization is for the `poktroll.session.MsgUpdateParam` message.
+		// However, this message is already authorized for the `pokt10d07y265gmmuvt4z0w9aw880jnsr700j8yv32t` address.
```
Comment (Olshansk) [inline app/upgrades/v0.0.11.go:66]: Is this the alpha or beta PNF address

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: app/upgrades/v0.0.11.go (hunk 3 above)
Comment (okdas) [reply app/upgrades/v0.0.11.go:26]: It should be an `-rc` here. We created RC release to bootstrap beta TestNet. The release must have this code (the upgrade handler) inside the binary. So we're just using the RC as a version to create a release on.

**Review by Olshansk — APPROVED** [review]

File: app/upgrades/v0.0.11.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+			err = keepers.SessionKeeper.SetParams(ctx, sessionParams)
+			if err != nil {
+				logger.Error("Failed to set session params", "error", err)
+				return err
+			}
+			logger.Info("Successfully updated session params", "new_params", sessionParams)
+
+			// Set tokenomics params. The values are based on default values for LocalNet/Beta TestNet.
+			// Validate with: `poktrolld q tokenomics params --node=https://testnet-validated-validator-rpc.poktroll.com/`
+			tokenomicsParams := keepers.TokenomicsKeeper.GetParams(ctx)
```
Comment (okdas) [inline app/upgrades/v0.0.11.go:45]: Here's the line that caused a panic: https://gist.github.com/okdas/50f417e3a86230e9e1aea85da841e1de

I need to investigate why I did not have this issue during local upgrade from v0.0.10 to v0.0.11. Will try to replicate using ONLY binaries provided in releases.

Could it be the before state is different on Alpha TestNet? If so, going to document this so we check for that in the future.

Comment (Olshansk) [reply app/upgrades/v0.0.11.go:45]: > Here's the line that caused a panic: [gist.github.com/okdas/50f417e3a86230e9e1aea85da841e1de](https://gist.github.com/okdas/50f417e3a86230e9e1aea85da841e1de)

Why did this cause the panic?

Comment (okdas) [reply app/upgrades/v0.0.11.go:45]: @Olshansk My first hypothesis it has happened due to the protobuf type change, but that should have been caught on the local test. For some reason, it did not. I'm going to be looking into that today.

I expect we will change an upgrade procedure to migrate each module individually, which can preserve the types between upgrades. There are some pros and cons associated with that. So far, I wanted the upgrade process to look simple, but it seems like additional complexity might worth it to avoid such issues in the future. 

Basically, I need to spend more time to provide a complete answer.

Comment (Olshansk) [reply app/upgrades/v0.0.11.go:45]: > I need to spend more time to provide a complete answer.

This is REALLY REALLY REALLY REALLY important, so please dive as deep as necessary to the finest grain of detail.

**Comment by okdas** [comment]
I'm going to remove the merge from main into this branch and do a force push.
This branch is not ready for code from main yet, as I'm still figuring out what happened on Alpha TestNet.

**Comment by Olshansk** [comment]
@okdas Following up on an offline discussion.

> This branch is not ready for code from main yet, as I'm still figuring out what happened on Alpha TestNet.

If you can't resolve it by mid-day today, let's do the following:

1. Post your latest update + investigation + details
2. Please move over to load-testing
3. @bryanchriswhite will pick it up next week
4. If @bryanchriswhite doesn't figure it out by mid week next week, I'll just in and help
5. @okdas Please be ready (aside from working on (2)) to potentially help out with some questions

**Review by Olshansk — CHANGES_REQUESTED** [review]
Undoing my approval per the latest discussion in this thread

**Comment by okdas** [comment]
Here's my investigation so far: https://www.notion.so/buildwithgrove/Alpha-TestNet-v0-0-11-Postmortem-152a36edfff680d58411c8fa384f877c?pvs=4

TLDR: KVStore had a different representation of empty params for the tokenomics module (alpha testnet vs local node), which explains the difference in behavior. 

I've also included some action items for us (me mostly) to do to avoid this and issues that were uncovered during this upgrade / postmortem investigation.

@bryanchriswhite I would love you to look into this as well, and I'm happy to pair/help.

**Comment by okdas** [comment]
OK, and now that I sent the previous message, it struck me what exactly happened here.

Alpha TestNet was provisioned using `v0.0.9`, which had a different protobuf for the `tokenomics` module params: https://github.com/pokt-network/poktroll/blob/v0.0.9/proto/poktroll/tokenomics/params.proto#L16

And it had a value, too: https://github.com/pokt-network/pocket-network-genesis/blob/master/shannon/testnet-alpha/genesis.json#L1018

That protobuf was then removed in `v0.0.10`, but the bytes were still preserved in the KVStore: `082A` - [see them decoded here](https://gchq.github.io/CyberChef/#recipe=From_Hex('Auto')Protobuf_Decode('',false,false)JSON_Minify()&input=MDgyQQ).

As I did not see any changes in the protobuf when comparing `v0.0.10` to `v0.0.11`, I deemed it safe to read the parameters from KVstore, while it was not.

If we did not remove the parameter or used a different field ID for the new fields, this wouldn't happen.

**Comment by bryanchriswhite** [comment]
**TL;DR, we SHOULD use `reserved` for deleted fields AND we SHOULD be versioning our protobuf types (e.g. `proto/tokenomics/v1beta1`).**

> OK, and now that I sent the previous message, it struck me what exactly happened here.

:exploding_head: Nice sleuthing! :rocket: 

> If we did not remove the parameter or used a different field ID for the new fields, this wouldn't happen.

Point taken that going forward, we need to be much more careful about backwards compatibility. In this particular case, I don' think there's any way we could've avoided removing this field as it was necessary to resolve a dependency cycle. In cases like this, it seems like the correct thing to do, going forward, would be to remove the field but also add its field number (and possibly field name) as [`reserved`](https://protobuf.dev/programming-guides/proto3/#reserved-field-numbers).

FWIW, this case would also be a bit unusual, IMHO, to resolve with versioned types as we would end up having multiple branches in the effected module's `InitGenesis` function to handle the two different versions of genesis. I think this makes much less sense then the typical case, which would be where a message or query handler has conditional branches for each version. "Genesis backwards-compatibility" seems like a bit of an oxymoron. :joy: 

This (re-)raises the point of **versoined protobuf source**, which we currently are not setup for, neither from a go package nor a proto source package persective. I think we will need to do something like the following in order to pre-allocate sufficient degrees of freedom to ensure we can safely increment the protobuf types into the future, without causing caos from both the go module/pkg and buf.build perspectives:

### Current proto directory structure:
```
./proto
└── poktroll
    ├── application
    │   └── *.proto
    │   └── *.proto
    ├── gateway
    │   └── *.proto
    ├── proof
    │   └── *.proto
    ├── service
    │   └── *.proto
    ├── session
    │   └── *.proto
    ├── shared
    │   └── *.proto
    ├── supplier
    │   └── *.proto
    └── tokenomics
    │   └── *.proto
```

### Conventional (versioned):
```
./proto
└── poktroll
    ├── application
    │   └── v1beta1
    │       └── *.proto
    ├── gateway
    │   └── v1beta1
    │       └── *.proto
    ├── proof
    │   └── v1beta1
    │       └── *.proto
    ├── service
    │   └── v1beta1
    │       └── *.proto
    ├── session
    │   └── v1beta1
    │       └── *.proto
    ├── shared
    │   └── v1beta1
    │       └── *.proto
    ├── supplier
    │   └── v1beta1
    │       └── *.proto
    └── tokenomics
        └── v1beta1
            └── *.proto
```

**Comment by Olshansk** [comment]
Firstly, great investigation @okdas!

@bryanchriswhite I want to avoid the `module/v1beta1/*.proto` refactor just now.

Let's continue with:
- Only incrementing IDs
- Using `reserved fields` for deprecated fields

Things we'll look out for:
- How do other cosmos chains do this?
- When (if) we hit major proto refactors, we'll revisit package versioning

**Comment by okdas** [comment]
I'll have to force push to remove the merge.
Screenshot 2025-01-23 at 4 20 06 PM

We should only let commits until v0.0.11. Please don't update the branch.

**Comment by okdas** [comment]
Now that Alpha TestNet is on `v0.0.11` - I merged main and this should be good to merge.

File: app/upgrades/v0.0.11.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+// Beta TestNet was launched with v0.0.11, so this upgrade is exclusively for Alpha TestNet.
+//   - Before: v0.0.10
+//   - After: v0.0.11
+var Upgrade_0_0_11 = Upgrade{
+	PlanName: "v0.0.11",
+	CreateUpgradeHandler: func(mm *module.Manager,
+		keepers *keepers.Keepers,
+		configurator module.Configurator,
+	) upgradetypes.UpgradeHandler {
+
```
Comment (Olshansk) [inline app/upgrades/v0.0.11.go:25]:
```suggestion
```

File: app/upgrades/v0.0.11.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+		// Adds new parameters using ignite's config.yml as a reference. Assuming we don't need any other parameters.
```
Comment (Olshansk) [inline app/upgrades/v0.0.11.go:26]: Is this comment still correct?

File: app/upgrades/v0.0.11.go [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+		// https://github.com/pokt-network/poktroll/compare/v0.0.10...v0.0.11-rc
```
Comment (Olshansk) [inline app/upgrades/v0.0.11.go:27]: Is this still an rc?

File: app/upgrades/v0.0.11.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+		applyNewParameters := func(ctx context.Context) (err error) {
+			logger := cosmosTypes.UnwrapSDKContext(ctx).Logger()
+			logger.Info("Starting parameter updates for v0.0.11")
+
+			// Set num_suppliers_per_session to 15
+			// Validate with: `poktrolld q session params --node=https://testnet-validated-validator-rpc.poktroll.com/`
```
Comment (Olshansk) [inline app/upgrades/v0.0.11.go:33]: Is this the correct URL and value has been validated?

**Review by Olshansk — APPROVED** [review]
Approving to unblock you, but just leaving a few comments for you to take into consideration.

File: app/upgrades/v0.0.11.go (hunk 8 above)
Comment (okdas) [reply app/upgrades/v0.0.11.go:26]: Yes

File: app/upgrades/v0.0.11.go (hunk 9 above)
Comment (okdas) [reply app/upgrades/v0.0.11.go:27]: Yes, and always will be. We needed the diff between v0.0.10 and v0.0.11-rc. This is a historical code. We don't even need it in the main branch. But I want to preserve to use as a reference.

File: app/upgrades/v0.0.11.go (hunk 10 above)
Comment (okdas) [reply app/upgrades/v0.0.11.go:33]: Yes and yes.
